enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+related ``DNSName`` CRs will be executed. This might add some delay to the DNS lookup process. However, this will only
+happen whenever there's a change in the DNS information.
+
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+* This enhancement will be tested through e2e tests by adding EgressFirewall rules containing regular DNS names
+and wildcard DNS names.
+* Testing the feature where IPs are dynamically changed may be a little bit tricky as this will probably include
+creation of DNS records and then changing the IP adresses of the DNS records through the e2e tests.
+
+### Graduation Criteria
+
+This is a user facing change and will directly go to GA. This feature requires an update to Openshift Docs.
+
+#### Dev Preview -> Tech Preview
+
+N.A. This feature will go directly to GA.
+
+#### Tech Preview -> GA
+
+N.A. This feature will go directly to GA.
+
+#### Removing a deprecated feature
+
+
+### Upgrade / Downgrade Strategy
+
+Upgrade expectations:
+* On upgrade, the OVN-K master will create the corresponding ``DNSName`` CRs for each DNS name in the
+existing EgressFirewall resources. The ``egressfirewall`` plugin will also start updating the ``.status``
+fields of the ``DNSName`` CRs. The scenarios arising out of the order of the update of the various components
+are dicussed in [Version Skew Strategy](#version-skew-strategy)
+
+Downgrade expectations:
+* On downgrade, the ``DNSName`` CRs may still remain. However, these CRs would not have any impact on how
+EgressFirewall ACL rules are implemented in the downgraded cluster. Deleting the CR Definition of ``DNSName``
+from the cluster would remove all the ``DNSName`` CRs.
+
+### Version Skew Strategy
+
+The following 2 scenarios may occur during the upgrade process:
+* Scenario 1: The Cluster DNS operator and the CoreDNS pods are upgraded first and then the OVN-K master pods.
+
+  In this scenario, the ``egressfirewall`` CoreDNS plugin will start inspecting each DNS lookup before the ``DNSName``
+  CRs are created by the OVN-K master. The plugin will just respond with the response received from other plugins for
+  the DNS lookups. As OVN-K master will be continuing the DNS lookups for DNS names with expired TTLs, CoreDNS will
+  also be responding with the corresponding IPs and the TTLs.
+
+* Scenario 2: The OVN-K master pods are upgraded first and then the Cluster DNS operator and the CoreDNS pods.
+
+  In this scenario, the OVN-K master will create ``DNSName`` CRs for each unique DNS name used in EgressFirewall rules.
+  However, as the Cluster DNS operator and the CoreDNS pods are still not upgraded, CoreDNS pods will not run the
+  ``egressfirewall`` plugin. The OVN-K master will still receive the response for the DNS lookup queries it will send
+  to the CoreDNS.
+
+### Operational Aspects of API Extensions
+
+
+#### Failure Modes
+
+
+#### Support Procedures
+
+
+## Implementation History
+
+
+## Alternatives
```

Comment:
I think we probably need to fill in this section, because as the discussion goes on, it turns out that (surprise!) this solution, like the others proposed, has some annoying problems too...

Alternatives:
- The existing system (explain how it works, and its pros and cons)
- Tell users to use a SOCKS or HTTP proxy instead (where the proxy does DNS and can be configured to allow or reject arbitrary DNS names). This would work _much_ better for the cases where it works, but it's much less transparent for the clients. But it's _SO_ much better that it might be worth explicitly pointing this out in our docs that admins should consider if this would work for them.
- Actually modify the DNS response rather than just snooping it; if the client requests a denied DNS name, they just get back `REFUSED`. People suggest this idea a lot as being the "obvious" implementation, but it really only makes sense if you are doing "deny *.example.com", which (I have argued), you should not be doing. In order to support it for the "allow *.example.com" case, then that implies that we would have to block all DNS requests that are not specifically allowed, but no one would expect that saying "allow *.example.com" (or even "deny 0.0.0.0/0") would block all other DNS. So I don't think this is really an option. (Also, it has the problem that if the pod has access to DNS via any other means, then they can subvert your policy.)
- Anything else? What do Calico and Cilium do?

Reply:
https://github.com/freedge/dnsflow !?

Reply:
In Cilium the IP information corresponding to a DNS name is captured through a DNS Proxy which runs on each Cilium agent: https://docs.cilium.io/en/latest/security/policy/language/#dns-based

Reply:
So, Cilium does both. It re-resolves the DNS names periodically as well as uses a DNS Proxy to get the response of DNS queries to get the IP information corresponding to the DNS names.

Reply:
> https://github.com/freedge/dnsflow !?

This solution also is kind of similar to that of Cilium. Instead of running a DNS Proxy on each agent, a separate daemonset is deployed which tracks the pods for each specific node. The daemoset also gets the DNS traffic through DNSTap plugin and extracts the response of DNS queries for the pods running on the specific node. It then updates the OVS flows directly of the node.

Reply:
In another comment thread on this PR, danw mentioned the idea of using gRPC instead of CRs, which seems like an alternative worth mentioning.